accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+  that check for breakpoint/watchpoint status.
+
+* We will need to either maintain a mapping from locals to register/stack slot
+  for all Wasm PC points that we might hit a breakpoint or watchpoint, or force
+  winch to unconditionally spill locals to the stack. The latter would greatly
+  simplify tracking local locations, while the former would greatly increase the
+  amount of context we would need to pass into the utility function for
+  inspecting the execution state.
+
+# Future Work
+
+## Source Language Debugging
+
+While DWARF metadata is enough to enable debugging compiled languages, it
+doesn’t really help with debugging a running interpreter that has been compiled
+to Wasm — debugging that wasm module would amount to debugging the interpreter,
+not the underlying program being interpreted. Let’s examine this through two
+examples: debugging a Rust program compiled with DWARF annotations, and
+debugging a JavaScript programming being interpreted by SpiderMonkey compiled to
+Wasm.
+
+In the case of debugging compiled Rust that has DWARF annotations present, all
+of the existing workflows we described in this document would apply: source line
+mappings and the locations of local variables would all be determined by
+translating directly back to offsets in the wasm module. However if we’re
+debugging javascript running on SpiderMonkey compiled to Wasm, DWARF annotations
+would only give us information about the spidermonkey interpreter; we need to
+instruct the debugging infrastructure provided by SpiderMonkey to proceed with
+source-level debugging.
+
+The difference between these two workflows suggests that a debugger
+implementation in Wasmtime needs to understand the target being debugged at a
+deeper level than just annotations. To address this problem, we plan to enrich
+the debug adapter interface described for the null and DWARF adapters above with
+enough information to allow control of a guest interpreter. This will include
+the need to reflect on symbols exported by the interpreter allowing for
+functions to be dynamically called by the debug adapter, reusing all of the
+interpreter’s existing debugging facilities directly.
+
+## Record + Replay
+
+Once we have both core dumps and the live debugging capabilities described
+above, we start to open the door to more interesting debugging capabilities. One
+of those is the ability to record a trace in a production setting, and then
+replay it in a debug environment to aid with the quick reproduction of a bug.
+This is a major feature of the [rr](https://rr-project.org/) debugger, which is
+able to record the results of syscalls and deterministically replay program
+execution with that trace to enable easy debugging of otherwise
+tricky-to-reproduce bugs. As WebAssembly execution is already well-sandboxed and
+semi-deterministic, recording the result of calling imports would enable the
+same sort of offline analysis of a production failure.
```

Comment:
I'm really excited for this!

![back to the future time circuits](https://media2.giphy.com/media/v1.Y2lkPTc5MGI3NjExZmo0YTBtbDNwZnlvbDY0bmxrMnE2MHVjMXdibzR6NWdhampuNWRzayZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/xFPPCepF6cMkO4Ie9k/giphy.gif)